src/ray/core_worker/transport/raylet_transport.cc

```diff
@@ -77,9 +97,83 @@ void CoreWorkerRayletTaskReceiver::HandleAssignTask(
   // rpc reply first before the NotifyUnblocked message arrives,
   // as they use different connections, the `TaskDone` message is sent
   // to raylet via the same connection so the order is guaranteed.
-  RAY_UNUSED(raylet_client_->TaskDone());
+  RAY_UNUSED(raylet_client_->TaskDone(num_tasks_completed));
   // Send rpc reply.
-  send_reply_callback(status, nullptr, nullptr);
+  send_reply_callback(Status::OK(), nullptr, nullptr);
+}
+
+Status CoreWorkerRayletTaskReceiver::HandleAssignTask0(
+    const rpc::AssignTaskRequest &request, const TaskSpecification &task_spec) {
+  // Set the resource IDs for this task.
+  ResourceMappingType resource_ids;
+  auto resource_infos =
+      flatbuffers::GetRoot<protocol::ResourceIdSetInfos>(request.resource_ids().data())
+          ->resource_infos();
+  for (size_t i = 0; i < resource_infos->size(); ++i) {
+    auto const &fractional_resource_ids = resource_infos->Get(i);
+    auto &acquired_resources =
+        resource_ids[string_from_flatbuf(*fractional_resource_ids->resource_name())];
+
+    size_t num_resource_ids = fractional_resource_ids->resource_ids()->size();
+    size_t num_resource_fractions = fractional_resource_ids->resource_fractions()->size();
+    RAY_CHECK(num_resource_ids == num_resource_fractions);
+    RAY_CHECK(num_resource_ids > 0);
+    for (size_t j = 0; j < num_resource_ids; ++j) {
+      int64_t resource_id = fractional_resource_ids->resource_ids()->Get(j);
+      double resource_fraction = fractional_resource_ids->resource_fractions()->Get(j);
+      if (num_resource_ids > 1) {
+        int64_t whole_fraction = resource_fraction;
+        RAY_CHECK(whole_fraction == resource_fraction);
+      }
+      acquired_resources.push_back(std::make_pair(resource_id, resource_fraction));
+    }
+  }
+
+  std::vector<std::shared_ptr<RayObject>> results;
+  auto status = task_handler_(task_spec, resource_ids, &results);
+
+  auto num_returns = task_spec.NumReturns();
+  if (task_spec.IsActorCreationTask() || task_spec.IsActorTask()) {
+    RAY_CHECK(num_returns > 0);
+    // Decrease to account for the dummy object id.
+    num_returns--;
+  }
+
+  RAY_LOG(DEBUG) << "Assigned task " << task_spec.TaskId()
+                 << " finished execution. num_returns: " << num_returns;
+  if (results.size() != 0) {
+    RAY_CHECK(results.size() == num_returns);
+    for (size_t i = 0; i < num_returns; i++) {
+      ObjectID id = ObjectID::ForTaskReturn(
+          task_spec.TaskId(), /*index=*/i + 1,
+          /*transport_type=*/static_cast<int>(TaskTransportType::RAYLET));
+      Status status = object_interface_.Put(*results[i], id);
```

Comment:
I think you want to use `PutAsync` here to match the `FlushAsyncPuts` called in `ProcessAssignedTasks`?